### PR TITLE
Binding properties

### PR DESCRIPTION
**Pull request by dconnolly**
Included an overview of binding properties for KEMs and for ML-KEM specifically.

@bwesterb @sophieschmieg might have opinions to improve the text

### CODE REVIEW

**Comment by dconnolly** [comment]
> I do not believe that ML-KEM suffers from reencapsulation attacks. As such, I do not see the point in including an extended paragraph on them. Remember, this is not about KEM's in general, but ML-KEM specifically. A more targeted paragraph (which mentions only those concerns specific to ML-KEM would be more reasonable.

The mention is a motivation for binding properties in general, and an earlier version of Signal PQXDH using a regular IND-CCA KEM was vulnerable to such an attack - they were saved by using an earlier version of Kyber, in part, but had to commit to more information in their AAD to mitigate it completely. They are relevant in how the recommended use of ML-KEM as now standardized in FIPS 203 is safe, but using another KEM in a similar way may not be (because IND-CCA is insufficient). We can pare it down though.

**Comment by sfluhrer** [comment]
The other issue is that this is mainly advice for protocol designers (and to a lesser extend, implementors).  If there is something they may need to be aware of (e.g. MAL-BIND-K-KT), we should explain what it is (in addition to pointing them to some academic paper they will have trouble understanding), when that threat is relevant, and how to mitigate that threat.

**Comment by dconnolly** [comment]
> The other issue is that this is mainly advice for protocol designers (and to a lesser extend, implementors). If there is something they may need to be aware of (e.g. MAL-BIND-K-KT), we should explain what it is (in addition to pointing them to some academic paper they will have trouble understanding), when that threat is relevant, and how to mitigate that threat.

Yes, that is basically done in the ML-KEM-specific section at the end

File: draft-sfluhrer-cfrg-ml-kem-security-considerations.md [hunk 1]
```diff
@@ -310,22 +382,46 @@ Alice).  The cryptographical libraries that Alice and Bob use may
...
-enough of the secret that successive queries determine the private
+key generation is performed and expanded. Lattice public keys are a lattice
+and a secret hidden by an error term; if additional error can be introduced
+into the public key generation stage, then the success of decapsulation can
+reveal enough of the secret that successive queries determine the private
 key. Notably, this means a public key can be 'poisoned' such that a future
 adversary can recover the private key even though it will appear correct in
 normal usage.
 
+Per the analysis of the final {{FIPS 203}} in Sophie Schmieg’s [KEMMY24], a
```
Comment (danvangeest) [inline draft-sfluhrer-cfrg-ml-kem-security-considerations.md:401]:
```suggestion
Per the analysis of the final {{FIPS203}} in Sophie Schmieg’s [KEMMY24], a
```

Comment (dconnolly) [reply draft-sfluhrer-cfrg-ml-kem-security-considerations.md:401]: Ah thank you, whoops

**Comment by sfluhrer** [comment]
I've been going through it, and it's hard to come up with a plausible scenario where either MAL-BIND-K-CT nor MAL-BIND-K-PK is a threat (even in the case where ML-KEM is used for authentication). Both attack models assume that the adversary is either the private key holder, or can inject bogus private keying material).  However if they can pick the private key, well, they can already do pretty much what they want anyways (and so you don't have security in any case).

If we can't think of cases where these security properties might be needed, I don't know if they need to be mentioned.  Remember, this is not an academic paper - this tries to instruct protocol designers and implementors about things they need to keep in mind, and there's little point in having them worry about things that aren't actually a concern.

Am I missing something?

File: draft-sfluhrer-cfrg-ml-kem-security-considerations.md [hunk 2]
```diff
@@ -96,35 +123,46 @@ sends the ciphertext to Alice, who uses her private key to generate the same
...
+means that ML-KEM is secure to use with static public keys.
+
+# Properties beyond IND-CCA
+
+There are security properties beyond IND-CCA such as those studied in [CAS].
+There are many variants. MAL-X is hardest to achieve, but failure to achieve
+it hasn't lead to practical attacks at present. LEAK-B is in the middle, and
+failure to be LEAK-X has lead to reencapsulation attacks [PQXDH]. ML-KEM achieves
+LEAK-X, but not all MAL-X. The latter is discussed in the 'Security Properties' 
+section below.
```
Comment (dconnolly) [inline draft-sfluhrer-cfrg-ml-kem-security-considerations.md:154]:
```suggestion
There are security properties beyond IND-CCA such as those studied in [CAS].
There are many variants. MAL-X is hardest to achieve, but failure to achieve
it hasn't lead to practical attacks at present. LEAK-X is in the middle, and
failure to be LEAK-X has lead to reencapsulation attacks [PQXDH]. ML-KEM achieves
LEAK-X, but not all MAL-X. The latter is discussed in the 'Security Properties' 
section below.
```

**Comment by sfluhrer** [comment]
This text is still too targeted towards academics (it doesn't even define what MAL-BIND-K-CT or MAL-BIND-K-PT is, and instead refers to some paper which nonacademics will not understand), and doesn't talk about when some protocol designer would need to be concerned.  I did add a paragraph to the main text trying to describe the situation (in terms that a protocol designer would understand) - you might want to review that.

Also, I have to wonder - if MAL-BIND-* assumes that the attacker has control of the private key, why should we assume that there's any security properties at all?

**Comment by sophieschmieg** [comment]
No disagreement with you or the decision to close, just as additional context: I did have a scenario where an attacker controlled private key having or not having certain properties mattered, although it was not MAL-BIND-* in that case:

The scenario was that the protocol used a implicitly authenticated key exchange, i.e. both parties know a static public key of the other party. It computed the shared secret as KDF(ECDH(static_Alice, static_Bob) || ECDH(ephemeral_Alice, ephemeral_Bob)). This is a valid, two side authenticated, forward secret key exchange. Another team wanted to have a one side authenticated version of this protocol, and they did this by using a shared static key for Alice. This would technically be only a "LEAK" scenario in the terms of Cas' paper, but you could imagine that shared static key sitting on a writeable network share. The problem is that the handshake above has key compromise impersonation, so instead of degenerating into a one-side authenticated key exchange if one private key is compromised, it degenerates into an unauthenticated key exchange, since Eve can impersonate Bob to Alice and compute the shared secret using Alice's static private key and her ephemeral private key. A 3DH handshake KDF(ECDH(static_Alice, ephemeral_Bob) || ECDH(ephemeral_Alice, static_Bob) || ECDH(ephemeral_Alice, ephemeral_Bob)) degenerates as expected in this case.

The other, similar scenario where attacker controlled secret keys play a role is in the case of invisible salamanders. You can formulate those as various AEADs, AES-GCM most of all, not having LEAK-BIND-CT-PT, i.e. an attacker with knowledge of the key can craft a ciphertext that will decrypt to a different plaintext under a different key. That property also has caused real world vulnerabilities, but it doesn't apply to implicitly rejecting KEMs to begin with, as they never bind the ciphertext to the shared secret.

So in the end, while there are relevant examples of vulnerabilities that require the attacker to know the private key, none of the examples I am aware of, even in hypothetical protocols designed only to be vulnerable to misbinding attacks require KEMs binding against maliciously formed keys. In a document aimed at developers in particular, it would be hard to convey these nuances in a non-academic fashion, and the advice more or less becomes "Have cryptographers look at your protocol, protocol design is hard and can introduce vulnerabilities even if secure primitives are used". Even a generic "Always hash in everything" imho doesn't quite capture the problems here, since it already presumes a specific use case (and one where you probably should use TLS or HPKE), and in ML-KEM's case does not even apply, as it has stronger binding guarantees than e.g. X25519 or RSA.

**Pull request closed without merging**